### ISSUE TITLE
asyncBuffer docs are misleading

### ISSUE THREAD

**Issue by conartist6** [iter-tools/iter-tools#425]
I made a mistake when I got rid of `asyncMapParallel` and other parallel functions. I thought I could instead have parallelism be created with `asyncBuffer`. `asyncBuffer` is still useful and has a correct implementation and example, however the docs suggest (as I was thinking) that the method can create parallelism, which it cannot. Making multiple requests to an async generator simply queues them.

Here is an example of code that I just wrote in macrome where I want parallelism and I believe I have implemented it correctly:

```js
const initialPaths = await traverse(this.projectRoot, { ignored });

const operations = await execPipe(
  initialPaths,
  filter((path) => !!this.generatorInstances.find(({ generator }) => matches(path, generator))),
  map((path) =>
    this.parserFor(path)
      .hasHeader(path)
      .then((h) => (h ? { path, operation: ADD } : null)),
  ),
  asyncBuffer(4),
  asyncFilter((op) => op !== null),
  arrayFromAsync,
);

this.initialProcess(operations);
```

**Comment by coder-mike** [comment]
I also came across this.

From what I can see, the docs are actually wrong. The example in docs today is this:

```js
pipe(
  range,
  asyncMap(i => fetch(`/page/${i}`).then(res => res.json()),
  asyncSlice(0, 50),
  asyncBuffer(4),
  asyncToArray
);
```

For one, there is a missing close-paren and `toArray` needs to be `asyncToArray` otherwise it doesn't run at all.

For another, if I make this example self contained, as in the following code, it doesn't actually "fetch 4 pages at once" as the documentation says:

```js
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

const operations = pipe(
  range,
  asyncMap(processItem),
  asyncSlice(0, 50),
  asyncBuffer(4),
  asyncToArray
);
operations();

async function processItem(item) {
  console.log(`Fetching page ${item}`);
  await delay(1000);
  console.log(`Page ${item} fetched`);
}
```

Outputs:
```
Fetching page 0
Page 0 fetched
Fetching page 1
Page 1 fetched
Fetching page 2
Page 2 fetched
Fetching page 3
Page 3 fetched
Fetching page 4
Page 4 fetched
Fetching page 5
...
```

(unless somehow I've messed up the example in trying to make it self-contained?)

And then this begs the question of how to actually achieve what the example is trying to do, which is to get a level of parallelism when pipelining a chain of async operations on an async (or sync) data source.

**Comment by coder-mike** [comment]
Looking at this further, I think I've found a work-around. Here is a self-contained example that actually achieves what the documentation appears to be trying to do:

```js
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

const operations = pipe(
  range,
  asyncMap(item => [processItem(item)]),
  asyncSlice(0, 50),
  asyncBuffer(4),
  asyncMap(([processing]) => processing),
  arrayFromAsync
);
operations();

async function processItem(item) {
  console.log(`Fetching page ${item}`);
  await delay(1000);
  console.log(`Page ${item} fetched`);
}
```

<details>
<summary>Output</summary>

```
Fetching page 0
Fetching page 1
Fetching page 2
Fetching page 3
Fetching page 4
Page 0 fetched
Fetching page 5
Page 1 fetched
Fetching page 6
Page 2 fetched
Fetching page 7
Page 3 fetched
Fetching page 8
Page 4 fetched
Fetching page 9
Page 5 fetched
Fetching page 10
Page 6 fetched
Page 7 fetched
Page 8 fetched
Page 9 fetched
Page 10 fetched
```

</details>

The key here is that the callback passed to `asyncMap` does not return a promise, and so the pipeline will not wait for it but instead continue with the next item. The callback instead returns a wrapped representation of the promise.

The newly added step `asyncMap(([processing]) => processing)` is the point in the pipeline when it will wait for each promise, but by this time it's already buffered 4 other promises in parallel.

The above example, as in the docs, consumes a _synchronous_ data source, but I think a more common requirement will be the processing of an asynchronous data source (e.g. values from a database or event stream). The above pattern generalizes to asynchronous data sources as well, as in the following complete example:

```js
const { execPipe, asyncMap, asyncConsume, asyncBuffer } = require('iter-tools');

const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

execPipe(
  dataSource(),
  asyncMap(i => [processItem(i)]),
  asyncBuffer(5),
  asyncMap(([promise]) => promise),
  asyncConsume
)

async function* dataSource() {
  for (let i = 0; i < 10; i++) {
    console.log(`Producing ${i}`)
    delay(100);
    yield i;
  }
}

async function processItem(item) {
  console.log(`Processing item ${item}`);
  await delay(1000);
  console.log(`Item ${item} processed`);
}
```

<details>
<summary>Output</summary>

```
Producing 0
Processing item 0
Producing 1
Processing item 1
Producing 2
Processing item 2
Producing 3
Processing item 3
Producing 4
Processing item 4
Producing 5
Processing item 5
Item 0 processed
Producing 6
Processing item 6
Item 1 processed
Producing 7
Processing item 7
Item 2 processed
Producing 8
Processing item 8
Item 3 processed
Producing 9
Processing item 9
Item 4 processed
Item 5 processed
Item 6 processed
Item 7 processed
Item 8 processed
Item 9 processed
```

</details>

**Comment by conartist6** [comment]
I think the use case you have works already??? You would write it like this:

```js
execPipe(
  dataSource(),
  asyncBuffer(5),
  asyncMap(i => processItem(i)),
  asyncConsume
)
```

In that code `asyncBuffer` would continue making requests from `dataSource` even in the absence of pressure from `asyncConsume`, which in this case there is no pressure from consume since it is only requesting items as often as values from `processItem(i)` are available, so `asyncBuffer` really is creating pressure that causes buffering to occur. That is how this method is meant to work, and I believe it should. It looks to me like there's a problem in your implementation of `dataSource`: you forgot to `await` `delay()`.

**Comment by coder-mike** [comment]
> I think the use case you have works already???

@conartist6 Can you clarify what you mean by "works already"? 

The use case I have is that I want some level of parallelism when processing an asynchronous input stream. The example code I gave to address this use case is working code, yes.

The point I was making is that the documentation also has an example that purports to achieve something very similar but fails to do so correctly. 

For anyone coming to this thread for the same reason I came here, I believe that my code shows a way to solve the problem that the documented example code does not.

**Comment by conartist6** [comment]
No sorry I see, that doesn't help.

**Comment by conartist6** [comment]
I've thought about it a lot and my conclusion is this: I like your mechanism, but it isn't self-documenting enough to recommend as a pattern. If I saw that in code there's a 50% chance I'd rip it out the array wrapping before I came across the docs that explain why it does something in this particular circumstance. Also if the user needs to do an `asyncFilter` in parallel it's going to be even more complicated. For that reason I want to hide as much as possible of the magic of parallelization behind documented APIs, even if they're essentially doing your trick under the hood.

**Comment by conartist6** [comment]
I tidbit form my investigation. I thought we couldn't return a promise without wrapping it because we do `yield await func()`, but actually getting rid of the `await` does nothing. Here is an example showing that `yield` actually still unwraps the promise:
```js
const it = (async function* () {
  yield Promise.resolve(5);
})();
it.next().then(({value}) => console.log(value)); // 5, not Promise<5>
```

This would seem to mean that only your solution or something like it will work.

**Comment by conartist6** [comment]
I guess my thinking at the moment is that we really need `mapParallel`, `filterParallel`, and `flatMapParallel` back. I got rid of them because I thought this method could replace them, but I was wrong. My problem with all the other solutions is really just that they aren't self-contained. They create an intermediate state that only makes sense in the context of parallelization and that the user must remember to clean up correctly or else it will leak into other code. By making the APIs self-contained they become self-documenting and there is no longer a worry that they can leak internal representations of half-finished work into other transformations.

**Comment by conartist6** [comment]
If I do that I'll need to reconsider #224 which I had closed.

**Comment by coder-mike** [comment]
> but it isn't self-documenting enough to recommend as a pattern

Yeah, good point.

> I guess my thinking at the moment is that we really need mapParallel, filterParallel, and flatMapParallel back. I got rid of them because I thought this method could replace them

I like your idea of building these up from primitives instead of having them baked in.

Would it be possible to have something like:

```ts
function parallel<T, U>(
  numberToRunInParallel: number, 
  operationToRunInParallel: (input: AsyncIterable<T>) => AsyncIterable<U>,
  input: AsyncIterable<T>,
): AsyncIterable<U>;
```

With the intended usage something like this:

```js
  await execPipe(
    source(),
    asyncMap(stepA),
    asyncMap(stepB),
    parallel(2, pipe(
      asyncMap(stepC),
      asyncMap(stepD),
    )),
    asyncMap(stepE),
    asyncMap(stepF),
    asyncConsume
  )
```

Here, if we assume that `stepC` and `stepD` take longer for some reason, and we want to parallelize the work.

To give a crude illustration:

![IMG_2802](https://user-images.githubusercontent.com/2261100/126368549-a3e1dc74-91eb-499c-80f2-08703c319328.gif)

(here I'm assuming that steps C and D take 2 seconds each, and the rest take 1 second each)

The semantics could be deterministic and would allow parallelization of any other operation or sequence of operations.

What do you think?

**Comment by conartist6** [comment]
So you've introduced a third kind of parallelization I think. Here are the three kinds we have:

1. parallelizing fetching items from an async iterable while doing async processing on items
2. parallelizing multiple async callbacks on a sync or async data source
3. parallelizing multiple different kinds of processing

The solution for each is different. The first can be done with `asyncBuffer`. The second can be done with your array wrapping trick or `mapParallel`. The last looks like a use case for `fork`.

**Comment by conartist6** [comment]
I like your little animated illustration there but you're going to get tripped up by the pull-based mechanics. So yes, you could build what you've drawn and use `fork` to split and `roundRobin` to recombine. Let's game that out with `filter` instead of `map`:

The first value is requested. `roundRobin` is the last step, and thus is evaluated first. Its initial state causes it to request an item from the left fork. Let's say the source has three values in it and only the first should be filtered out. The left fork filters out the first value, but a value was requested so unless left is done it must still produce one. The second value is pulled from source and fork puts it in a buffer since it's for right not left. The third value is pulled form source and goes to the left filter, which it passes. If you've kept track of all this you see the problem: we've yielded the third value before the second value.

In essence the issue is that you can't parallelize without knowing anything about what operation you're trying to parallelize.

**Comment by conartist6** [comment]
I think having `flatMapParallel` is going to offer a lot of utility because it can act like a map and a filter at the same time, meaning you can put everything you want to parallelize into the same operation. For example to filter a value out, just return `[]`. To map, return `[fn(value)]`. And you can even replace one value with many by returning an iterable of values. If you have a series of operations you want parallelized, you could just chain them together inside `flatMapParallel`.

**Comment by coder-mike** [comment]
> 1. parallelizing fetching items from an async iterable while doing async processing on items ...  The first can be done with asyncBuffer.

A buffer sometimes solves this and sometimes doesn't. For example, if I define the source like this (which seems fairly natural to me) then the buffer doesn't cause multiple fetches to be done in parallel:

```js
async function* source() {
  while (someCondition) {
    yield fetchAnItemFromTheDatabase();
  }
}
```

A developer could just "know" this (that buffering has a different kind of effect on different input iterators), but I've been tripped up by this in the past.

That's not to say that buffering is useless on the above `source`. Here's an example where I think buffering helps, even with no parallelism:

```js
const source = async function* () {
  for (let i = 0; i < 3; i++){
    log(`Producing ${i}`)
    await delay(1000);
    yield i;
  }
}

const processItem = async function (i) {
  log(`Processing ${i}`);
  await delay(1000)
  log(`Processed ${i}`);
}

// Unbuffered
await execPipe(
  source(),
  asyncMap(processItem),
  asyncConsume
);
/*
    +1ms Producing 0
 +1005ms Processing 0
 +1010ms Processed 0
    +0ms Producing 1
 +1014ms Processing 1
 +1009ms Processed 1
    +0ms Producing 2
 +1014ms Processing 2
 +1003ms Processed 2
*/

// Buffered 
await execPipe(
  source(),
  asyncBuffer(1),
  asyncMap(processItem),
  asyncConsume
);
/*
    +1ms Producing 0
 +1010ms Producing 1
    +2ms Processing 0
 +1008ms Processed 0
    +1ms Producing 2
    +0ms Processing 1
 +1008ms Processed 1
    +0ms Processing 2
 +1003ms Processed 2
*/
```

![Photo 21-7-21, 08 33 08](https://user-images.githubusercontent.com/2261100/126404636-5bc058be-a652-436d-9ea9-76fcda48351f.gif)

The buffered example has twice the throughput at steady state, even when both the source and processing are strictly serial.

I'm not sure that I'm saying anything useful here. I guess my instinct is that it _may_ be possible to define a system that preserves the following invariants:

  1. That `async function* (source) { yield* source }` can be considered as the "identity operator" (since it certainly appears to be). That applying it to any iterator produces a new iterator that behaves the same way as the original. 
  2. It follows then that all iterators must be processed only serially (that consumers agree to call `next()` only after the successful completion of the previous).
  3. That a stream operator function (like `asyncMap`) either _always_ introduces parallelism (like `parallel()` or `flatMapParallel`) or it _never_ introduces parallelism (like `asyncMap`).

**Comment by coder-mike** [comment]
> The first value is requested. roundRobin is the last step, and thus is evaluated first. Its initial state causes it to request an item from the left fork. Let's say the source has three values in it and only the first should be filtered out. The left fork filters out the first value, but a value was requested so unless left is done it must still produce one. The second value is pulled from source and fork puts it in a buffer since it's for right not left. The third value is pulled form source and goes to the left filter, which it passes. If you've kept track of all this you see the problem: we've yielded the third value before the second value.

Yes, there's an interesting problem here. I wonder if it's a fundamental problem or just a problem with the implementation we're imagining.

It may be fundamental. The parallel streams need to be recombined, and there is not necessarily a way of defining a deterministic ordering between the outputs since they're not necessarily related to particular inputs. A pathological case would be a transform that didn't even use its input and instead just generated output on a timer.

So, yeah, I agree that `flatMapParallel` makes the most sense.

**Comment by conartist6** [comment]
Invariant 3 is already pretty much a guarantee in the iterator spec. Any iterators implemented with generator functions will force effectively serial calls to `next()`. Generators do this by buffering the calls to `next()`. That is one thing that makes it difficult to come up with inventive solutions. I really do not want to break with generator semantics.

**Comment by conartist6** [comment]
Tangent here, but what are you using the library for? I'm always curious since I don't get feedback any other way.

**Comment by coder-mike** [comment]
> Tangent here, but what are you using the library for? I'm always curious since I don't get feedback any other way.

I've used twice, in completely different projects. 

The first one was an alerting engine. 

 - Given a sequence of events from another server, determine whether there's an issue that needs to be raised as an alert (e.g. email someone). 
 - Needed to define various guarantees, such as "if the server goes down in the middle of processing an event, it should continue where it left off". This was implemented using a persistent bookmark/cursor that recorded the last successfully processed event. Storing the bookmark is itself an asynchronous process.
 - For performance reasons, it would have been crazy to process one event before requesting the next (the data source had hundreds of events per second), so we needed to have a processing pipeline that could process hundreds of events in parallel if necessary.

The second use case (which I'm working on now) is polling a legacy SQL database for new records in a particular table and then acting on them in an asynchronous manner. 

 - The records are being fetched in batches using an async generator which performs a round-trip to the database. 
 - This is followed by an asyncBuffer so that we're requesting the next batch while processing the previous (we don't need batches to be requested in parallel, but just need the next batch to be "ready for us" when we're finished processing the previous). 
 - For each record, we're contacting a web server, which is a round-trip that we don't want to wait for so this requires parallelism in the pipeline
 - And again, we need to keep track of the "progress" through the records so that even if the server dies in the middle, it's still guaranteed to process each record at least once. So there is a round-trip database update at the end of each item

**Comment by conartist6** [comment]
I guess you're not blocked on any of this since we have workarounds for everything. At the moment my main goal is finishing pulling out the build system I created into its own package. Once that is done my plan is to my plan is to do a bit of reorganization of the imports (I want `import {map} from 'iter-tools/async'`) and then put out the next major version and start trying in earnest to gain adoption. As far as I know I'm quite a ways behind the other major libraries in usages.

**Comment by coder-mike** [comment]
> I guess you're not blocked on any of this since we have workarounds

Yep :-)

> At the moment my main goal is

Awesome. Lots of good stuff to look forward to! Thanks for your great contribution to the ecosystem.

**Comment by richardscarrott** [comment]
> Looking at this further, I think I've found a work-around. Here is a self-contained example that actually achieves what the documentation appears to be trying to do:
> 
> ```js
> const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
> 
> const operations = pipe(
>   range,
>   asyncMap(item => [processItem(item)]),
>   asyncSlice(0, 50),
>   asyncBuffer(4),
>   asyncMap(([processing]) => processing),
>   arrayFromAsync
> );
> operations();
> 
> async function processItem(item) {
>   console.log(`Fetching page ${item}`);
>   await delay(1000);
>   console.log(`Page ${item} fetched`);
> }
> ```
> The key here is that the callback passed to `asyncMap` does not return a promise, and so the pipeline will not wait for it but instead continue with the next item. The callback instead returns a wrapped representation of the promise.
> 
> The newly added step `asyncMap(([processing]) => processing)` is the point in the pipeline when it will wait for each promise, but by this time it's already buffered 4 other promises in parallel.
> 
> The above example, as in the docs, consumes a _synchronous_ data source, but I think a more common requirement will be the processing of an asynchronous data source (e.g. values from a database or event stream). The above pattern generalizes to asynchronous data sources as well, as in the following complete example:
> 
> ```js
> const { execPipe, asyncMap, asyncConsume, asyncBuffer } = require('iter-tools');
> 
> const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
> 
> execPipe(
>   dataSource(),
>   asyncMap(i => [processItem(i)]),
>   asyncBuffer(5),
>   asyncMap(([promise]) => promise),
>   asyncConsume
> )
> 
> async function* dataSource() {
>   for (let i = 0; i < 10; i++) {
>     console.log(`Producing ${i}`)
>     delay(100);
>     yield i;
>   }
> }
> 
> async function processItem(item) {
>   console.log(`Processing item ${item}`);
>   await delay(1000);
>   console.log(`Item ${item} processed`);
> }
> ``` 

@coder-mike are you sure the above pattern generalizes to asynchronous data sources as, if you add the missing `await` to the dataSource delay it doesn't in fact produce anything in paralell e.g.

```js
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

execPipe(
  dataSource(),
  asyncMap((i) => [processItem(i)]),
  asyncBuffer(5),
  asyncMap(([promise]) => promise),
  asyncConsume
);

async function* dataSource() {
  for (let i = 0; i < 10; i++) {
    console.log(`Producing ${i}`);
    await delay(1000); // Added missing `await` and made delay more noticeable
    yield i;
  }
}

async function processItem(item) {
  console.log(`Processing item ${item}`);
  await delay(1000);
  console.log(`Item ${item} processed`);
}
```

However, perhaps this is down to the way async generators work where multiple reads will just queue the results, e.g.

```js
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

async function* dataSource() {
  for (let i = 0; i < 10; i++) {
    await delay(1000);
    yield i;
  }
}

const src = dataSource();
console.time("read");
await Promise.all([src.next(), src.next(), src.next()]);
console.timeEnd("read");
// read: 3s (not 1s) 
```

But maybe other async data sources such as Node streams do perform parallel requests when next is called 🤔

UPDATE: I just tested Node streams and they also don't perform requests in parallel when consumed as an iterable, e.g.

```js
import * as stream from "stream";

const dataSource = () => {
  let i = 0;
  return new stream.Readable({
    highWaterMark: 0,
    read() {
      console.log(`Producing ${i}`);
      setTimeout(() => {
        this.push(i.toString());
        if (i === 10) {
          this.push(null);
          return;
        }
        i++;
      }, 1000);
    },
  });
};

const src = dataSource()[Symbol.asyncIterator]();
console.log(src);
console.time("read");
await Promise.all([src.next(), src.next(), src.next()]);
console.timeEnd("read");
// read: 3s (not 1s) 
```

**Comment by conartist6** [comment]
Yep, the biggest problem with this behavior has been the way async generators queue requests. Calling `next(); next(); next()` doesn't actually create any pressure on an async generator.

You're on the right track in realizing that you need a sync generator of promises, not an async generator.

**Comment by richardscarrott** [comment]
@conartist6 interesting I hadn't thought of that; so an "async" data source is more powerful if written using a sync generator?

```js
function* dataSource() {
  for (let i = 0; i < 10; i++) {
    yield new Promise(async (resolve) => {
      console.log(`Producing ${i}`);
      let _i = i;
      await delay(1000);
      resolve(_i);
    });
  }
}

const src = dataSource();
console.time("read");
await Promise.all([src.next().value, src.next().value, src.next().value]);
console.timeEnd("read");
// read: 1s!
```

I guess then you could buffer the promises which would control the read parallelism?

```js
execPipe(
  dataSource(),
  asyncBuffer(4),
  asyncTap((i) => {
    console.log("tap", i);
  }),
  asyncConsume
);
```

Aside, I basically ignored iterators for years because I couldn't see how they could compete with Node stream performance and frankly I don't want to write a for loop (🤷‍♂️) which 90% of resources start and end with, so it's really cool to see this library, really good work! Now just trying to figure out the exact patterns to replace my Node stream workflows.

**Comment by conartist6** [comment]
I really like for loops actually. I'm not really ready to start recommending this as an alternative to streams until we get some of this perf stuff worked out, but I'm hoping it'll happen.

**Comment by conartist6** [comment]
It still hurts my brain a bit to think about. The sync data source of promises can be buffered, yes, but you also have to think about unhandled rejections then. You also have to be careful because a yielded promise will still be awaited in a `for await .. of` loop.

**Comment by richardscarrott** [comment]
yeah I need to get my head around the idea a bit more / test out some patterns; I guess I really mean to say I don't want to write a for loop because it pushes you so hard into writing serial pipelines.

**Comment by felixfbecker** [comment]
Sorry for jumping into this old thread, but is there a conclusion on how to currently best parallelize a map operation, other than returning promises wrapped in another object?

**Comment by conartist6** [comment]
I have not been thinking about this issue. AFAIK the best solution is some variant of using a wrapper object. Maybe we could make some nice APIs around that?

**Comment by felixfbecker** [comment]
It doesn't look so bad in the inline-pipeline example above, but the concern I have to use the wrapper object is that it makes for hard-to-understand APIs when the map is inside a re-usable function, because then the wrapper object becomes part of the API contract of that function.


Example:

```ts
// I have to explain the weird API here and hope consumers will understand what to do:
/**
 * Produces items.
 *
 * @return An async iterable of array wrappers of promises. This is so that the caller can control parallelism with asyncBuffer(). The caller must destructure the array and resolve the promises with asyncMap().
 */
function produce(): AsyncIterable<[Promise<Item>]> {
  return execPipe(source(), asyncMap(item => [fetchData(item)]))
}

// Consumer now needs to know about the weird array wrapper and flatten it
for await (const item of execPipe(produce(), asyncBuffer(5), asyncMap(([promise]) => promise))) {
  console.log(item)
}
```


The use case of using AsyncIterables and parallelism is usually because I want the _caller_ of a producer to be able to control parallelism, while the producer does the map operation (avoiding the model of `p-map` where the producer needs to also decide the parallelism). So it is common to have the "function boundary" exactly there between the mapper and the buffer.

It would be a lot more ergonomic to have the `parallelMap()` helper in this case so a simple `AsyncIterable<Item>` can be returned that can still be piped to `asyncBuffer()` without any extra transformation needed.

**Comment by conartist6** [comment]
Another interesting part of this is that we're basically running up against a limitation in the design of generator functions and for/await loops. In the time since I did this work the official iterator helpers proposal has made some progress, and I believe they're currently thinking about the same [issues](https://github.com/tc39/proposal-async-iterator-helpers/issues).

**Comment by conartist6** [comment]
@felixfbecker in the example you shared you don't need the wrapper trick at all. You synchronously know all the fetches that need to be made, and so there is then no conflict with the sync implementation of `map` which passes Promises through untouched. Where it gets complicated is if you want to go async -> async with some parallelism.

**Comment by felixfbecker** [comment]
Yeah sorry, you're right, I meant to use `asyncMap()`. Updated.

If the synchronous `map()` was used, you'd still get an awkward return type like `Iterable<Promise<T>>` – what I'd want for an ergonomic API is an `AsyncIterable<T>`.

I also realized the limitation when implementing AsyncIterables as generators, but I think there would still be value in a solution for parallelism especially for `map` operations, so you can at least have parallelism as long as you express the parallizable parts of an operation using operators like `map`/`mapParallel` instead of `for` loops + `yield`.

**Comment by conartist6** [comment]
I'm just not completely convinced that it's possible/something you should want to try to do.

When you have `Promise<{done, value}>` as in an async iterable things start to get dangerous mighty quick because a reordering of items could mean that the sequence terminator arrives before some of the sequence values

**Comment by conartist6** [comment]
What does it *mean* to parallelize pull based streaming? I cannot answer that question in simple words, and I have not seen anyone else give a succinct and convincing answer either.

To me these parallelization questions are interesting, but they simply aren't pull-based streaming. They necessarily break down the basic lazy patterning of pull-based streaming which in which control flows from the sink to the source. If the sink isn't controlling execution of the source (e.g. handling exceptions) aren't you really just talking about push-based streaming?

**Comment by conartist6** [comment]
I guess I'm back to where I started, which is that the only viable solution is wrappers. Could they be abstracted behind some kind of API that would save you always having to write out `asyncMap(([promise]) => promise)`? I assume so.

But they still create this basic problem where structured exception handling breaks down. As soon as you put that fetch promise into an array instead of merging it into your current thread of control you lose the ability to know whether its acceptance or rejection will ever really be handled. Most likely a rejection would be an unhandled rejection, which would crash the runtime (if it can). You have to set up a promise rejection handler by the time the promise rejects to avoid that, and that would be a race condition even with the wrapper-based code.

**Comment by conartist6** [comment]
@felixfbecker Looping you in on [this thread I made to sort out my thoughts](https://github.com/tc39/proposal-async-iterator-helpers/issues/11). It seems that there is a way to design the helpers that allows backpressure to propagate better, where in my design it tends to be dispersed by generator function semantics.

**Comment by felixfbecker** [comment]
I think the take-away is:

- Only _generators_ don't support parallelism, but general AsyncIterators (and therefore operators) can, so operators like `map` in `iter-tools` have to be implemented without generators.
- Errors can still be propagated as expected (i.e. immediate abort of the chain, just like `Promise.all()` or `Promise.race()`) as long as the operator implementation forwards them correctly.

**Comment by conartist6** [comment]
That sounds accurate. It'll be a breaking change though so there's a good bit of work that needs to happen before a changed implementation would be out in the world. It would certainly light a bit of a fire under my butt if the async helper proposals moved forward, but mostly I'm working on my extensible streaming parser framework/LSP competitor.

**Comment by anacierdem** [comment]
@conartist6 Can you elaborate on this a little:
> As soon as you put that fetch promise into an array instead of merging it into your current thread of control you lose the ability to know whether its acceptance or rejection will ever really be handled

Do you mean that it is possible to *forget* to follow it up with another step that will unwrap it from the array and add to the flow or is it something else? I think once you put a promise back into the chain, it wouldn't cause an undhandled exception.

Working on something similar BTW, this is what I have come up with instead:

```js
await execPipe(
  dataSource(),
  // Create a sequence of iterators of size N, i.e the batch size
  asyncBatch(N),
  // Pull a group of N and cache it in an array
  asyncMap(arrayFromAsync), 
  asyncFlatMap((batch) => batch.map(processItem)),
  asyncConsume
);
```

This is essentially the same technique as above in that it still wraps them with an array but keeps the batch in the same array and I think it is a little bit more readable. Still it is not as clean as I would want to and there is a quirk that I am still investigating. For some reason it pulls one more time (for N+1) from the source before starting the batch.

**Comment by conartist6** [comment]
@anacierdem Yep that's a pattern that I'm aware of, and is the best you can do with `iter-tools` right now. It's funny that you bring that up because I'm embroiled in huge discussion about exactly this right now over on the [TC39 async iterators proposal](https://github.com/tc39/proposal-async-iterator-helpers/issues/20). I am really hoping that it's possible to make something slightly better, because it seems to me like a lot of people are going to want something like this

**Comment by anacierdem** [comment]
The other issue I am having right now looks something like this:

```js
const source = () => {
  return {
    count: 0,
    async next() {
      console.log('next');
      await delay(1000);
      return { value: this.count++, done: false };
    },
    [Symbol.asyncIterator]() {
      return this;
    },
  };
};

const iter = execPipe(
  source(),
  asyncBatch(2),
  asyncMap(arrayFromAsync),
  asyncMap((batch) => {
    console.log('will start batch');
  })
);

console.log(await iter.next());
```

Note that I am only pulling a single time from the final iterator but the output looks like this:

```
next
next
next
will start batch
```

This means that before starting the batch operation, we are waiting for another value from source. I was hoping to add a buffer in between and overlap the batch request with the 3rd pull from the source but there is something going on here that I couln't yet figure out.

**Comment by conartist6** [comment]
Yep, correct, that's the whole cleverness of the pattern right there, what I am calling "backpressure". The idea is to ensure that the producer always is ready to do more work because it knows the consumer has requested it.

Normally the producer would produce some data and then halt because only after the client processed that data would the client come back and request more. The idea is to eliminate the delay in production by having the client make requests for the next several pieces of data all at once.

**Comment by anacierdem** [comment]
I am not on board yet. The consumer only asked for a single item here. The intermediery iterator (via `asyncBatch`) should pull 2 items from the underlying source to return an iterator of size 2 to the downstream consumer. But it instead pulls 3 instead of 2.

This is in fact introducing additional delay. Consider the source takes 1000ms for each item. In this case to be able to start the batch, we need to wait for 3s instead of 2, increasing the total duration.

Edit: Note that I don't have an `asyncBuffer` in the above example yet.

Edit 2: I think in the long run (and in practice) this will just average out especially when used in tandem with `asyncBatch` but this is still surprising to me.

Edit 3: Related blogpost [here](https://alinacierdem.com/async-iterators-the-obvious-secret/).

**Comment by conartist6** [comment]
Now you're talking about what sounds like a bug in the implementation of `batch`. I would strongly recommend you file that bug with as much of repro as you can work out.

My point was that you probably want something more like this:

```
const iter = execPipe(
  source(),
  asyncBuffer(2),
  asyncMap((item, i) => {
    console.log(`processing item ${i}`);
  })
);
```

`asyncBatch` does not create backpressure, which means that as you say it basically just costs additional delay. It requests, waits for a response, requests, waits for a response. Therefore with `asyncBatch` there will be no speedup associated with increasing the batch size. 

`asyncBuffer` does not work the same way: buffer makes requests for items but then instead of waiting for the response before requesting more items, it just makes more requests. When implemented correctly your source would see `next` `next` `next` *immediately* (syncrhonously) and only after synchronously making those backpressure requests would the consumer stop and wait for the first response

**Comment by anacierdem** [comment]
I will report it with as much info I can extract 👍🏼 `asyncBuffer` is indeed queues multiple pulls that will run in parallel with whatever async operation you have afterwards. That is pretty nice, I agree.

**Comment by conartist6** [comment]
The main problem is that my async helpers prevent backpressure from propagating through them properly because they are implemented as async generators, which means that they inherently allow pressure but do not do anything with it. In other words, `asyncMap(asyncBuffer(source))` may be an effective optimization, but in iter-tools right now `asyncBuffer(asyncMap(source))` would not be.

At this point it seems likely that this limitation will be dropped in the next major release of iter-tools. TC39 has chosen to implement their async `map` in a way that does propagate backpressure so I will have to implement the feature as well to stay intercompatible and state-of-the-art. I just haven't started doing the work yet as the focus of my efforts is elsewhere.

**Comment by anacierdem** [comment]
Ah, I get it better now. That's the reason why I need to have some work _later_ in the pipeline with the current structure. Then, if I'm reading this right, it means an outer `.next()` call will propagate back to the original source when it is executed with the new structure. So whenever someone tries to do `next`, it will propagate back until every buffer is full. That will indeed enable much deeper optimizations.

Edit: I think this also means that we don't actually need this batch + array map trick if we could simply do `asyncBuffer(asyncMap(source))` and `asyncMap` could run those `next`s immediately rather than being in an async generator

**Comment by conartist6** [comment]
That's just what I've been saying over on the TC39 thread!

**Comment by Inviz** [comment]
Hello @conartist6 I read many of your posts here and elsewhere. It was a doozey of a journey, and i switched like 3 libraries - each sharing the same limitation. In the end I rolled my own: 

https://github.com/Inviz/iterators

It supports unordered concurrency, buffering and back-pressure over any kind of iterables, including dealing with async generators of async generators for batching

It has `map(fn, concurrency?)` and `concat(streams, concurrency?, bufferSize?)` basic functions that allow concurrently pulling promises from iterator without resolving them immediately. Internally it has a pub-sub mechanism, where buffered upstream values are matched against async consumers. This mechanism allows blocking both publishers (if buffer is full) and consumers (if buffer is empty). This allows racing values from streams, which is super useful for batch operations where order doesnt particularily matter. I do not have concurrency with strict order as of now, but there could be ways to emulate it.  

Biggest inspiration for actually solving the problem was https://www.npmjs.com/package/tiny-async-pool Literally it was the _only_ function that approaches to solving concurrency the right way. However it's too minimalistic and uses recursive Promise.race which isnt suited for large numbers of computations (causing high GC pressure), so i had to unroll it into pub-sub thing. 

Just contributing some ideas back, as you left some trace on the iternet across those years, thought you might be interested. I have no objection against reuse of code or concepts. Anybody is free to plunder  if they are so inclined.

(discussion: https://github.com/somnicattus/rotery/issues/3#issuecomment-2848867205)
